components/board/genericlinux/hw_pwm.go

```diff
@@ -0,0 +1,174 @@
+// Package genericlinux is for Linux boards. This particular file is for using sysfs to
+// interact with PWM devices. All of these functions are idempotent: you can double-export a pin or
+// double-close it with no problems.
+package genericlinux
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+type pwmDevice struct {
+	// These values are immutable
+	chipPath string
+	line     int
+
+	mu sync.Mutex
+
+	// These values are mutable
+	periodNs         uint64
+	activeDurationNs uint64
+	isExported       bool
+	isEnabled        bool
+}
+
+func newPwmDevice(chipName string, line int) (*pwmDevice, error) {
+	// There should be a single directory within /sys/devices/platform/<chipName>/pwm/, whose name
+	// is mirrored in /sys/class/pwm. That's the one we want to use.
+	chipDir := fmt.Sprintf("/sys/devices/platform/%s/pwm", chipName)
+	files, err := os.ReadDir(chipDir)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, file := range files {
+		if strings.Contains(file.Name(), "pwmchip") && file.IsDir() {
+			chipPath := fmt.Sprintf("/sys/class/pwm/%s", file.Name())
+			return &pwmDevice{chipPath: chipPath, line: line}, nil
+		}
+	}
+	return nil, errors.Errorf("Could not find any PWM device with name %s", chipName)
+}
+
+func writeValue(filepath string, value uint64) error {
+	// The permissions on the file (the third argument) aren't important: if the file needs to be
+	// created, something has gone horribly wrong!
+	return os.WriteFile(filepath, []byte(fmt.Sprintf("%d", value)), 0o600)
+}
+
+func (pwm *pwmDevice) writeChip(filename string, value uint64) error {
+	return writeValue(fmt.Sprintf("%s/%s", pwm.chipPath, filename), value)
+}
+
+func (pwm *pwmDevice) writeLine(filename string, value uint64) error {
+	return writeValue(fmt.Sprintf("%s/pwm%d/%s", pwm.chipPath, pwm.line, filename), value)
+}
+
+// Export tells the OS that this pin is in use, and enables configuration via sysfs.
+func (pwm *pwmDevice) export() error {
+	if pwm.isExported {
+		return nil // Already exported
+	}
+	if err := pwm.writeChip("export", uint64(pwm.line)); err != nil {
+		return err
```

Comment:
Great idea! I'm now wrapping the errors with the chipPath, line, and the pseudofile written to. The one limitation is that this won't distinguish errors enabling the pin from ones disabling it, since those use the same pseudofile for different purposes. but I can't imagine errors disabling the pin, and I'd sure expect them all to be from enabling it.